### PR TITLE
Why you should always use access tokens to secure an API

### PR DESCRIPTION
**Pull request by mpaktiti**
New p2 doc

### CODE REVIEW

File: articles/api-auth/why-use-access-tokens-to-secure-apis.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+:::
+
+The `access_token` can be any type of token (not necessarily a JWT) and is meant for the API. Its purpose is to inform the API that the bearer of this token has been authorized to access the API and perform specific actions (as specified by the `scope` that has been granted). In the example we used earlier, after you authenticate, and provide your consent that the to-do application can have read/write access to your calendar, an `access_token` is sent from Google to the to-do application. Each time the to-do application wants to access your Google Calendar it will make a request to the Google Calendar API, using this `access_token` in an HTTP `Authorization` header.
+
+## How NOT to use each token
+
+Now that we saw what these tokens can be used for, let's see what they cannot be used for.
+
+- __An `access_token` should not be used for authentication__. The reason for that is that it holds no information about the user. It cannot tell us if the user has authenticated and when. That's why OAuth 2.0 cannot be used for authentication. Instead, OpenID Connect works on top of OAuth 2.0 and provides user authentication.
```
Comment (cabralmartin) [inline articles/api-auth/why-use-access-tokens-to-secure-apis.md]: maybe add a note saying that access_tokens should be treated as opaque by clients. clients should not attempt to decode them or depend on a particular access_token format.

File: articles/api-auth/why-use-access-tokens-to-secure-apis.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+- __An `id_token` should not be used for API access__. Each token contains information on the intended audience (recipient). According to the OpenID Connect specification, the audience (claim `aud`) of each `id_token` must be the `client_id` of the client making the authentication request. If it isn't you shouldn't trust the token. An API, on the other hand, expects a token with the audience set to the API's unique identifier. So unless you are in control of both the client and the API, sending an `id_token` to an API will not work.
```
Comment (cabralmartin) [inline articles/api-auth/why-use-access-tokens-to-secure-apis.md]: the biggest case against id_tokens for api access is that the id_token (since it is issued to a particular client) is signed with a secret that is known to the client. if an API were to accept such token, it has no way of knowing if the client modified the token (say by adding more scopes) and then signed it again.

File: articles/api-auth/why-use-access-tokens-to-secure-apis.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    "https://${account.namespace}/userinfo"
+  ],
+  "azp": "${account.clientId}",
+  "exp": 1489179954,
+  "iat": 1489143954,
+  "scope": "openid profile email address phone read:appointments email"
+}
+```
+
+This token is meant for __authorizing the user to the API__. As such, it is completely opaque to clients, meaning that clients should not care about the contents of this token. Note that the token does not contain any information about the user itself besides their ID (`sub` claim), it only contains authorization information about which actions the client is allowed to perform at the API (`scope` claim).
```
Comment (cabralmartin) [inline articles/api-auth/why-use-access-tokens-to-secure-apis.md]: excellent! you may disregard my previous note on the subject of opaque tokens or reinforce it at the token.

File: articles/api-auth/why-use-access-tokens-to-secure-apis.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+# Why you should always use access tokens to secure an API
+
+There is much confusion on the Web about the differences between the OpenID Connect and OAuth 2.0 specifications, and their respective tokens. As a result many developers publish insecure applications, compromising their users security. The contradicting implementations between identity providers do not help either.
+
+This article is an attempt to clear what is what and explain why you should always use an access token to secure an API, and never an ID token.
+
+## Two complementary specifications
+
+OpenID Connect is used for __user authentication__. It enables you, the user, to verify your identity and give some basic profile information, without sharing your credentials.
```
Comment (cabralmartin) [inline articles/api-auth/why-use-access-tokens-to-secure-apis.md]: OpendId is a specification that augments OAuth. Oauth only dealt with authorization and doesnt care about how the authorization server verifies who is the user (authentication). this was a problem because home grown solutions for this were created (e.g. facebook connect).
openid augmented oauth by specifying how the AS can authenticate users. as a result, the spec is everything oauth had + authentication (the collateral of this was the introduction of id_tokens).

**Comment by jaredhanson** [comment]
LGTM

File: articles/api-auth/why-use-access-tokens-to-secure-apis.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    "https://${account.namespace}/userinfo"
+  ],
+  "azp": "${account.clientId}",
+  "exp": 1489179954,
+  "iat": 1489143954,
+  "scope": "openid profile email address phone read:appointments email"
+}
+```
+
+This token is meant for __authorizing the user to the API__. As such, it is completely opaque to clients, meaning that a client should not care about the contents of this token, decode it or depend on a particular token format. Note that the token does not contain any information about the user itself besides their ID (`sub` claim), it only contains authorization information about which actions the client is allowed to perform at the API (`scope` claim).
```
Comment (twistedstream) [inline articles/api-auth/why-use-access-tokens-to-secure-apis.md:90]: I'm not sure saying that the access token is meant for "authorizing the user to the API" is accurate. Isn't it really about "authorizing the client to the API"? OAuth2 doesn't really concern itself with a user per se. It provides flows that allow users to grant access, but the access is being granted to the client, not themselves. User authorization is something that has already occurred within the Authorization Server.

cc @jaredhanson

Comment (jaredhanson) [reply articles/api-auth/why-use-access-tokens-to-secure-apis.md:90]: I understand what you are getting at here, and you are technically correct.  It's impossible, in the strictest sense, for a human user to interact _directly_ with an API.  There's always a software-based agent (aka, the client) operating on behalf of the user that allows interaction with the API.

Keeping things in layman's terms, I have no issue with the phrase "authorizing the user to the API" since its for all intents and purposes usually the desired outcome.

File: articles/api-auth/why-use-access-tokens-to-secure-apis.md (hunk 2 above)
Comment (ChristianCiach) [reply articles/api-auth/why-use-access-tokens-to-secure-apis.md]: > the id_token (since it is issued to a particular client) is signed with a secret that is known to the client

I don't think this is true. The id_token is signed using the private key of the IDP, so by validating the id_token (which is a JWT) you can validate that the token really was issued by the IDP, thus, the included information about the user is valid.

Comment (cabralmartin) [reply articles/api-auth/why-use-access-tokens-to-secure-apis.md]: @ChristianCiach It really depends on the signing algorithm. If the `id_token` is using a symmetric key algorithm (such as HS256) the key used is the client secret (this is [per spec](http://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation)). If the `id_token` is signed using one of the asymmetric key algorithms (e.g. RS* family) then Auth0 will use the tenant private key for signing it.
